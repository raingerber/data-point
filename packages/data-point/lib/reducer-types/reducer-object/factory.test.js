/* eslint-env jest */
'use strict'

const factory = require('./factory')
const createTransform = require('../reducer-expression').create

test('reducer/reducer-function#isObject', () => {
  expect(factory.isObject('$foo')).toBe(false)
  expect(factory.isObject(() => true)).toBe(false)
  expect(factory.isObject([])).toBe(false)
  expect(factory.isObject({})).toBe(true)
})

describe('reducer/reducer-function#getReducerProps', () => {
  it('should accept an empty object', () => {
    const reducer = factory.getReducerProps(createTransform, {})
    expect(reducer).toEqual([])
  })
  it('should accept a non-empty object', () => {
    const reducer = factory.getReducerProps(createTransform, {
      a1: '$a[]',
      b1: {
        a2: ['$a2', () => false],
        b2: [
          {
            a3: 'transform:entity-name',
            b3: '$b1.b2.b3'
          }
        ]
      }
    })
    expect(reducer).toMatchSnapshot()
  })
})

describe('reducer/reducer-path#create', () => {
  it('reducer object with no props argument', () => {
    const reducer = factory.create(createTransform)
    expect(reducer.type).toBe('ReducerObject')
    expect(reducer.props).toBeInstanceOf(Array)
  })
  it('reducer object with empty props argument', () => {
    const reducer = factory.create(createTransform, {})
    expect(reducer.type).toBe('ReducerObject')
    expect(reducer.props).toBeInstanceOf(Array)
  })
  it('reducer object with props argument', () => {
    const reducer = factory.create(createTransform, {
      a: '$a',
      b: '$b',
      c: '$c'
    })
    expect(reducer.type).toBe('ReducerObject')
    expect(reducer.props).toBeInstanceOf(Array)
    expect(reducer.props).toHaveLength(3)
  })
})
