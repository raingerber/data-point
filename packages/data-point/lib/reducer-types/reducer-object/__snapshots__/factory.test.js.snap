// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer/reducer-function#getReducerProps should accept a non-empty object 1`] = `
Array [
  Object {
    "path": Array [
      "a1",
    ],
    "transform": ReducerExpression {
      "reducers": Array [
        ReducerPath {
          "asCollection": true,
          "body": [Function],
          "name": "a",
          "type": "ReducerPath",
        },
      ],
      "type": "ReducerExpression",
    },
  },
  Object {
    "path": Array [
      "b1",
      "a2",
    ],
    "transform": ReducerExpression {
      "reducers": Array [
        ReducerPath {
          "asCollection": false,
          "body": [Function],
          "name": "a2",
          "type": "ReducerPath",
        },
        ReducerFunction {
          "body": [Function],
          "type": "ReducerFunction",
        },
      ],
      "type": "ReducerExpression",
    },
  },
  Object {
    "path": Array [
      "b1",
      "b2",
    ],
    "transform": ReducerExpression {
      "reducers": Array [
        ReducerObject {
          "props": Array [
            Object {
              "path": Array [
                "a3",
              ],
              "transform": ReducerExpression {
                "reducers": Array [
                  ReducerEntity {
                    "asCollection": false,
                    "entityType": "transform",
                    "name": "entity-name",
                    "type": "ReducerEntity",
                  },
                ],
                "type": "ReducerExpression",
              },
            },
            Object {
              "path": Array [
                "b3",
              ],
              "transform": ReducerExpression {
                "reducers": Array [
                  ReducerPath {
                    "asCollection": false,
                    "body": [Function],
                    "name": "b1.b2.b3",
                    "type": "ReducerPath",
                  },
                ],
                "type": "ReducerExpression",
              },
            },
          ],
          "type": "ReducerObject",
        },
      ],
      "type": "ReducerExpression",
    },
  },
]
`;
